File: python/p014a.py
#!/usr/bin/python

limit = 1000000
cache = {}
cache[1] = 1
result = [1, 1]

for i in range(1, limit):
    n, cnt = i, 0
    while n not in cache:
        if n % 2 == 0:
            n = n // 2
        else:
            n = (3 * n) + 1
        cnt += 1

    tmp = cache[n] + cnt
    cache[i] = tmp
    if tmp > result[1]:
        result = [i, cache[i]]

print(result)

File: rust/src/bin/p014a.rs

use std::collections::HashMap;

fn main() {

    let limit = 1000000;
    let mut cache: HashMap<usize, usize> = HashMap::new();
    cache.insert(1, 1);
    let mut result = [1, 1];

    for i in (1..limit) {
        let mut cnt = 0;
        let mut n = i;

        while !cache.contains_key(&n) {
            if n % 2 == 0 {
                n /= 2;
            } else {
                n = (n * 3) + 1;
            }
            cnt += 1;
        }

        let tmp = *cache.get(&n).unwrap() + cnt;
        cache.insert(i, tmp);
        if tmp > result[1] {
            result = [i, tmp];
        }

    }

    println!("{} {}", result[0], result[1]);
}
